h2o-algos/src/main/java/hex/tree/SharedTreeModelWithContributions.java

```diff
@@ -53,14 +62,55 @@ public Frame scoreContributions(Frame frame, Key<Frame> destination_key, Job<Fra
             .outputFrame(destination_key, outputNames, null);
   }
 
+  @Override
+  public Frame scoreContributions(Frame frame, Key<Frame> destination_key, int topN, int topBottomN, boolean abs, Job<Frame> j) {
+    if (_output.nclasses() > 2) {
+      throw new UnsupportedOperationException(
+              "Calculating contributions is currently not supported for multinomial models.");
+    }
+
+    Frame adaptFrm = cleanFrame(frame);
+    final String[] contribNames = ArrayUtils.append(adaptFrm.names(), "BiasTerm");
+
+    int topNAdjusted = contributionComposer.checkAndAdjustInput(topN, adaptFrm.names().length);
+    int topBottomNAdjusted = contributionComposer.checkAndAdjustInput(topBottomN, adaptFrm.names().length);
+
+    int outputSize = Math.min((topNAdjusted+topBottomNAdjusted)*2, adaptFrm.names().length*2);
+    String[] names = new String[outputSize];
+    byte[] types = new byte[outputSize];
+
+    for (int i = 0, topFeatureIterator = 1, bottomFeatureIterator = 1; i < outputSize; i+=2, topFeatureIterator++) {
+      if (topFeatureIterator <= topNAdjusted) {
+        names[i] = "top_feature_" + topFeatureIterator;
+        names[i+1] = "top_value_" + topFeatureIterator;
+      } else {
+        names[i] = "bottom_top_feature_" + bottomFeatureIterator;
+        names[i+1] = "bottom_top_value_" + bottomFeatureIterator;
+        bottomFeatureIterator++;
+      }
+      types[i] = Vec.T_STR;
```

Comment:
try to save memory by using categorical output

Reply:
Implemented